src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -29,28 +30,38 @@ class MessengerPass implements CompilerPassInterface
 {
     use PriorityTaggedServiceTrait;
 
-    private $messageBusService;
-    private $messageHandlerResolverService;
     private $handlerTag;
+    private $busTag;
+    private $senderTag;
+    private $receiverTag;
 
-    public function __construct(string $messageBusService = 'message_bus', string $messageHandlerResolverService = 'messenger.handler_resolver', string $handlerTag = 'messenger.message_handler')
+    public function __construct(string $handlerTag = 'messenger.message_handler', string $busTag = 'messenger.bus', string $senderTag = 'messenger.sender', string $receiverTag = 'messenger.receiver')
     {
-        $this->messageBusService = $messageBusService;
-        $this->messageHandlerResolverService = $messageHandlerResolverService;
         $this->handlerTag = $handlerTag;
+        $this->busTag = $busTag;
+        $this->senderTag = $senderTag;
+        $this->receiverTag = $receiverTag;
     }
 
     /**
      * {@inheritdoc}
      */
     public function process(ContainerBuilder $container)
     {
-        if (!$container->hasDefinition($this->messageBusService)) {
+        if (!$container->hasDefinition('messenger.handler_resolver')) {
             return;
         }
 
-        if (!$container->getParameter('kernel.debug') || !$container->hasAlias('logger')) {
-            $container->removeDefinition('messenger.middleware.debug.logging');
+        foreach ($container->findTaggedServiceIds($this->busTag) as $busId => $tags) {
+            if ($container->hasParameter($busMiddlewaresParameter = $busId.'.middlewares')) {
+                $this->registerBusMiddlewares($container, $busId, $container->getParameter($busMiddlewaresParameter));
+
+                $container->getParameterBag()->remove($busMiddlewaresParameter);
+            }
+
+            if ($container->hasDefinition('messenger.data_collector')) {
+                $this->registerBusToCollector($container, $busId, $tags[0]);
```

Comment:
should we iterate `$tags` as well or throw in case of >1?

Reply:
I don't believe there is a use-case for a bus to be tagged multiple times 🤔